incubator/hnc/pkg/reconcilers/object.go

```diff
@@ -504,58 +503,72 @@ func (r *ObjectReconciler) writeObject(ctx context.Context, log logr.Logger, ins
 	log.V(1).Info("Writing", "dst", inst.GetNamespace(), "origin", srcInst.GetNamespace())
 
 	var err error = nil
-	var op string
 	stats.WriteObject(inst.GroupVersionKind())
 	if exist {
 		err = r.Update(ctx, inst)
-		op = "update"
 	} else {
 		err = r.Create(ctx, inst)
-		op = "create"
 	}
 	if err != nil {
-		r.setErrorConditions(log, srcInst, inst, op, err)
-		log.Error(err, "Couldn't write", "object", inst)
-	} else {
-		// Add the object to the map if it does not exist because we are confident that the object was updated/created
-		// successfully on the apiserver.
-		r.recordPropagatedObject(log, inst.GetNamespace(), inst.GetName())
+		// Don't log the error since controller-runtime will do it for us
+		return err
 	}
-	return err
+
+	// Add the object to the map if it does not exist because we are confident that the object was updated/created
+	// successfully on the apiserver.
+	r.recordPropagatedObject(log, inst.GetNamespace(), inst.GetName())
+	return nil
 }
 
-func (r *ObjectReconciler) setErrorConditions(log logr.Logger, srcInst, inst *unstructured.Unstructured, op string, err error) {
+// setConditions is called when the reconciler has performed all necessary actions and knows if
+// they've succeeded or failed. It re-locks the forest just long enough to set or clear all
+// conditions. Since we didn't hold the lock while we were contacting the apiserver, we need to be
+// aware that the hierarchy may have changed in arbitrary ways; see below for details.
+//
+// This function also enqueues any modified namespaces for reconciliation.
+func (r *ObjectReconciler) setConditions(log logr.Logger, srcInst, inst *unstructured.Unstructured, act syncAction, err error) {
 	r.Forest.Lock()
 	defer r.Forest.Unlock()
+	ao := api.NewAffectedObject(inst.GetObjectKind().GroupVersionKind(), inst.GetNamespace(), inst.GetName())
+	ns := r.Forest.Get(inst.GetNamespace())
 
-	ao := getAO(inst)
-	msg := fmt.Sprintf("Could not %s: %s", op, err.Error())
-	r.setCondition(log, ao, api.CannotUpdate, inst.GetNamespace(), msg)
-	if srcInst != nil {
-		r.setCondition(log, ao, api.CannotPropagate, srcInst.GetNamespace(), msg)
-	}
-}
-
-func (r *ObjectReconciler) setCondition(log logr.Logger, ao api.AffectedObject, code api.Code, nnm, msg string) {
-	r.Forest.Get(nnm).SetCondition(ao, code, msg)
-	r.enqueueNamespace(log, nnm, "Set condition "+string(code))
-}
+	switch {
+	case hasFinalizers(inst):
+		// Propagated objects can never have finalizers
+		if ns.SetCondition(ao, api.CannotPropagate, "Objects with finalizers cannot be propagated") {
+			r.enqueueNamespace(log, ns.Name(), "Set CannotPropagate since it has finalizers")
+		}
 
-func getAO(inst *unstructured.Unstructured) api.AffectedObject {
-	gvk := inst.GetObjectKind().GroupVersionKind()
-	return api.NewAffectedObject(gvk, inst.GetNamespace(), inst.GetName())
-}
+	case err != nil:
+		// There was an error updating this object
+		msg := fmt.Sprintf("Could not %s: %s", act, err.Error())
```

Comment:
The msg may not look right for `actionUnknown syncAction = ""`?

BTW Question: when would `actionUnknown` happen?

Reply:
It should never happen :) I'll make that a bit more clear elsewhere.